Userland/Applications/PDFViewer/PDFViewerWidget.cpp

```diff
@@ -38,11 +44,144 @@ void PDFViewerWidget::initialize_menubar(GUI::Menubar& menubar)
+    });
+    m_go_to_prev_page_action->set_enabled(false);
+
+    m_go_to_next_page_action = GUI::Action::create("Go to &Next Page", Gfx::Bitmap::load_from_file("/res/icons/16x16/go-down.png"), [&](auto&) {
+        if (m_viewer->current_page() >= m_viewer->document()->get_page_count() - 1)
+            return;
+
+        auto new_page_index = m_viewer->current_page() + 1;
+        m_viewer->set_current_page(new_page_index);
+        m_viewer->update();
+        m_page_text_box->set_text(String::number(new_page_index + 1));
+    });
+    m_go_to_next_page_action->set_enabled(false);
+
+    toolbar.add_action(*m_go_to_prev_page_action);
+    toolbar.add_action(*m_go_to_next_page_action);
+
+    m_page_text_box = toolbar.add<GUI::TextBox>();
+    m_page_text_box->set_enabled(false);
+    m_page_text_box->set_fixed_width(30);
+
+    m_page_text_box->on_up_pressed = [&] {
+        auto current_index_opt = m_page_text_box->text().to_int();
+        if (!current_index_opt.has_value()) {
+            m_page_text_box->set_text("1");
+            return;
+        }
+
+        auto current_index = current_index_opt.value() - 1;
+        if (static_cast<u32>(current_index) >= m_viewer->document()->get_page_count() - 1) {
+            m_page_text_box->set_text(String::number(m_viewer->document()->get_page_count()));
+            return;
+        }
+
+        current_index++;
+        m_page_text_box->set_text(String::number(current_index + 1));
+        m_viewer->set_current_page(current_index);
+        m_viewer->update();
+    };
+
+    m_page_text_box->on_down_pressed = [&] {
+        auto current_index_opt = m_page_text_box->text().to_int();
+        if (!current_index_opt.has_value()) {
+            m_page_text_box->set_text("1");
+            return;
+        }
+
+        auto current_index = current_index_opt.value() - 1;
+        if (current_index <= 1) {
+            m_page_text_box->set_text("1");
+            return;
+        }
+
+        current_index--;
+        m_page_text_box->set_text(String::number(current_index + 1));
+        m_viewer->set_current_page(current_index);
+        m_viewer->update();
+    };
+
+    auto on_focus_out = [&] {
+        u32 new_page_index;
+
+        auto current_index_opt = m_page_text_box->text().to_int();
+        if (!current_index_opt.has_value()) {
+            new_page_index = 0;
+            m_page_text_box->set_text("1");
+        } else {
+            auto current_index = current_index_opt.value();
+            if (current_index <= 0) {
+                new_page_index = 0;
+                m_page_text_box->set_text("1");
+            } else {
+                new_page_index = current_index;
+            }
+        }
+
+        m_viewer->set_current_page(new_page_index);
+        m_viewer->update();
+    };
+
+    m_page_text_box->on_focusout = [&] { on_focus_out(); };
+    m_page_text_box->on_return_pressed = [&] { on_focus_out(); };
+    m_page_text_box->on_escape_pressed = [&] { on_focus_out(); };
```

Comment:
Is not being able to assign the same lambda to multiple `Function`s intentional? It seems kinda of weird to me, but I don't really know how `Function` works.

Reply:
Yes, that would be copying the functor, which is almost never safe.